### DISCUSSION TITLE
Why behavioral programming for UI development

### DISCUSSION THREAD

**Discussion by EdwardIrby** [plaited/plaited#53] [Show and tell]
_The following is copied from our [playbook](https://github.com/plaited/plaited/blob/main/playbook/coding/learn-about-behavioral-programming.md) we'll be using our [@plaited/behavioral](https://www.npmjs.com/package/@plaited/behavioral) library to demonstrate usage_

## What is behavioral programming?[^1]

Behavioral Programming, or bProgram for short, is based on scenarios. Behavioral
programs are composed of threads of behavior. These threads run in parallel to
coordinate a sequence of events via a synchronized protocol. During a bProgram's
execution, when a thread wants to synchronize with its peers, it submits a
statement to the central event arbiter using the sync statement. This statement
is returned when we invoke a bProgram. A **sync** statement declares which
events the **thread** requests, which events it waits for (but does not
request), and which events it would like to block (that is, prevent from being
selected). Request, block and waitFor events can be described by passing a
RuleSet object to the sync statement as parameter.

```ts
export type RuleSet<
  T extends (Detail) = (Detail),
> = {
  waitFor?: ParameterIdiom<T> | ParameterIdiom<T>[];
  request?: RequestIdiom<T> | RequestIdiom<T>[];
  block?: ParameterIdiom<T> | ParameterIdiom<T>[];
};
```

After calling sync, the thread is blocked, waiting for the rest of the threads
to synchronize as well. When all threads have submitted their statements, the
arbiter selects an event that was requested but not blocked. It then wakes up
the threads that requested or waited for that event. The rest of the threads
remain blocked, until an event they requested or waited for is selected.

Behavioral Programming was introduced by Harel, Marron and Weiss in 2012 in a
[paper](../assets/behavioral-programming.pdf) published at the Communications of
the ACM.

To learn more watch
[Rethinking Software Systems: A friendly introduction to Behavioral Programming by Michael Bar Sinai, 2018](https://youtu.be/PW8VdWA0UcA)

### tldr;

- Since each thread can not only specify which event to wait for, which event to
  request and which event to block the state of the program becomes implicit.
- Any thread can block execution of other threads thus allowing us to
  effectively modifying program behavior by simply adding new threads that
  append to event log.
- This idiomatic approach to programming aligns perfectly with microinteractions

### Intuitions

> 1. UI is a function of data
> 2. Data is derived from an event log
> 3. Deciding when and what to append to the event log is the most complex part
   > of development
> 4. Finding more natural ways for deciding when and what to append to the event
   > log is the direction we should be moving towards

[^2]

With Behavioral Programming we control the event log using behavioral threads.
The key idea of behavioral programming is that we can create new threads that
change the behavior of already written threads. Thanks to this idea, when our
software specification changes, we don’t necessarily need to modify existing
code. We can add new threads that modify the behavior of other threads. Each
thread lives on its own and is unaware of other threads, but they’re all
interwoven at runtime, allowing them to interact with each-other in a very
[novel way](#what-is-behavioral-programming1).

What's truly novel about this approach to application development is that we are
no longer managing state explicitly. It is an implicit byproduct of the
behavioral threads we create which control the event log. We can even create
snapshots of the state at every synchronization point of a bProgram run, to see
how our application state changes as events flow through our bProgram.

#### Scenario

We want to create an app that controls hot and cold water taps, whose output
flows are mixed.

```ts
import { expect, test } from "bun:test";
import { bProgram, DevCallback } from "@plaited/behavioral";

test("logging", () => {
  const logs: Parameters<DevCallback>[0][] = [];
  /**
   * Initiate our bProgram and destructure behavioral
   * programming utility functions
   */
  const {
    /** adds behavioral threads to behavioral program  **/
    addThreads,
    /**
     * creates a behavioral thread from synchronization sets and/or other
     * behavioral threads
     */
    thread,
    /**
     * A behavioral thread that loops infinitely or until some
     * callback returns false.This function returns a threads
     */
    loop,
    /**
     * At synchronization points, each behavioral thread
     * specifies three sets of events:
     * 1. requested events: the thread proposes that these be
     * considered for triggering
     * 2. waitFor events: the thread asks to be notified when
     * any of them is triggered
     * 3. blocked events: the threads currently forbids
     * triggering any of these events
     */
    sync,
    /** trigger the run of the behavioral program by requesting
     * the event passed as an argument
     */
    trigger,
    /** connect an action callback to the behavioral program that is
     * called when request event type of the same name as our
     * callback is selected by our behavioral program's
     * central event arbiter
     */
    feedback,
  } = bProgram({
    /**
     * Example of logging callback we can pass to our bProgram
     * options dev field.
     */
    dev: (msg) => logs.push(msg), // Logging callback
  });

  /** Add our rules for bProgram execution */
  addThreads({
    addHot: thread(
      sync({ request: { type: "hot" } }),
      sync({ request: { type: "hot" } }),
      sync({ request: { type: "hot" } }),
    ),
    addCold: thread(
      sync({ request: { type: "cold" } }),
      sync({ request: { type: "cold" } }),
      sync({ request: { type: "cold" } }),
    ),
    mixHotCold: loop([
      sync({
        waitFor: { type: "hot" },
        block: { type: "cold" },
      }),
      sync({
        waitFor: { type: "cold" },
        block: { type: "hot" },
      }),
    ]),
  });

  const actual: string[] = [];
  /** Add action callback to our feedback function */
  feedback({
    hot() {
      actual.push("hot");
    },
    cold() {
      actual.push("cold");
    },
  });
  trigger({ type: "start" });
  expect(actual).toEqual([
    "hot",
    "cold",
    "hot",
    "cold",
    "hot",
    "cold",
  ]);
  expect(logs).toMatchSnapshot();
});
```

The snapshot from the resulting test gives us insight into the behavior and
state of the app temporally.

```js
const logs = [
  [
    {
      "priority": 0,
      "request": [
        {
          "detail": undefined,
          "type": "start",
        },
      ],
      "selected": "start",
      "thread": "start",
      "trigger": true,
      "waitFor": [
        {
          "cb": [Function],
          "type": "",
        },
      ],
    },
    {
      "priority": 1,
      "request": [
        {
          "type": "hot",
        },
      ],
      "thread": "addHot",
    },
    {
      "priority": 2,
      "request": [
        {
          "type": "cold",
        },
      ],
      "thread": "addCold",
    },
    {
      "block": [
        {
          "type": "cold",
        },
      ],
      "priority": 3,
      "thread": "mixHotCold",
      "waitFor": [
        {
          "type": "hot",
        },
      ],
    },
  ],
  [
    {
      "priority": 1,
      "request": [
        {
          "type": "hot",
        },
      ],
      "selected": "hot",
      "thread": "addHot",
    },
    {
      "priority": 2,
      "request": [
        {
          "type": "cold",
        },
      ],
      "thread": "addCold",
    },
    {
      "block": [
        {
          "type": "cold",
        },
      ],
      "priority": 3,
      "thread": "mixHotCold",
      "waitFor": [
        {
          "type": "hot",
        },
      ],
    },
  ],
  [
    {
      "priority": 1,
      "request": [
        {
          "type": "hot",
        },
      ],
      "thread": "addHot",
    },
    {
      "priority": 2,
      "request": [
        {
          "type": "cold",
        },
      ],
      "selected": "cold",
      "thread": "addCold",
    },
    {
      "block": [
        {
          "type": "hot",
        },
      ],
      "priority": 3,
      "thread": "mixHotCold",
      "waitFor": [
        {
          "type": "cold",
        },
      ],
    },
  ],
  ...
  /**
   * full logs @
   * https://github.com/plaited/plaited/blob/main/libs/behavioral/src/tests/__snapshots__/water.spec.ts.snap
   */ 
];
```

### Microinteractions and behavioral programming go together, like peanut butter and jelly

Modern web application are often the result of
[component driven user interface practices](). If a web app interface component
is even remotely interactive it has one or more [microinteractions]().

Dan Saffer’s model of microinteractions is a design concept that focuses on the
small, often overlooked details in product design that significantly influence
user experience. These minute interactions, such as the sound a device makes
when turned on or the animation of a loading icon, are fundamental to creating a
seamless, intuitive user experience.

<img src="https://raw.githubusercontent.com/plaited/plaited/main/playbook/assets/microinteraction_diagram.png" alt="Diagram of Dan Saffer’s model of microinteraction" width="450"/>

The model consists of four parts: trigger, rules, feedback, and loops & modes.
The trigger initiates the microinteraction, rules determine how it functions,
feedback that lets the user know what's happening, and loops & modes that govern
its larger behavior.

Saffer’s model emphasizes that well-designed microinteractions are engaging,
informative, and human-centered, often defining the overall feel and function of
a product.

Coincidentally this model align with behavioral programming well and
[@plaited/behavioral](https://www.npmjs.com/package/@plaited/behavioral) has
been designed specifically to work within this pattern.

#### 1. Trigger function

Our trigger function triggers an event on component.

#### 2. Thread function

Our thread function serves as mechanism for writing rules on how and if we
respond to an event on our component.

#### 3. Feedback function

Our feedback function is used to respond to requested events allowed by our
behavioral threads, the rules, and let component users know what's happening
with our interactive component.

#### 4. Loop function

The loop function is ideal for web interfaces where trigger feedback pairs can
be triggered often infinitely such as checking and unchecking a checkbox for
example.

```ts
type Loop = (
  rules: RulesFunc<any>[],
  condition?: () => boolean,
) => RulesFunc<any>;
```

As we can see in the type above the loop function takes an optional second
argument, a callback function that returns true or false. By passing a callback
function that gets the current mode of our component (i.e. checked or unchecked)
in enables us to control which rules we're applying at a larger level each time
our behavioral program is run.

---

[^1]: [What is Behavioral Programming?](https://bpjs.readthedocs.io/en/latest/BPjsTutorial/index.html)

[^2]: [On User Interface Development: appending to the event log](https://medium.com/@lmatteis/on-user-interface-development-appending-to-the-event-log-8d8ca966795d)

**Comment by michbarsinai** [comment]
Hi Edward,

* Behavioral Programming is commonly abbreviated as "BP". `BProgram` is an abbreviation for "behavioral program". In "classic" BP, a b-program is just a set of b-thread that synchronize with each other. In later BP systems (such as BPjs) a b-program can also have additional properties, such as synchronized storage, event selection strategy, and life-cycle listeners.

* Second paragraph: "The rest of the threads remain blocked" - note that "blocked" here has two meanings: the blocking events (BP semantics) and blocking a thread (regular concurrency semantics). I also had these issues initially, and people got confused, so today I'm using "suspended" for  b-threads and "blocked" for events.

* 2nd tl;dr point is a bit unclear. Blocking is indeed a unique feature in BP, and an important part of its power. But blocking does not append to the event log, rather it prevents events from being appended. I'm sure I've read this wrong, which is why I think it should be more clear.

* I love the intuitions part. Very good.

* Note on the parameters to waitFor and block: I see you use RequestIdiom in request, and ParameterIdiom in waitFor and block. I'm not sure what the difference is. In BPjs and BPjs, we use EventSets for block and waitFor. These are really just functions that get and event, and return true if the event is a member of the group, and false otherwise. This proved to be a rather powerful concept, since you can wait to "any event whose name starts with 'B'", for example. The trick is that an event is an event set that contains only itself, so passing an event to waitFor is just a private case of passing an event set.

* For the request part of the sync statement we need to use actual events, since these have to be selected and propagated to the b-threads. There's a concept around that in addition to event is should be possible to support event generators. I haven't heard anyone implementing it, though.

I've written about the structure of BPjs here ([https://slideshare.net/MichaelBarSinai/deep-dive-into-bpjs](https://t.co/ft3YQYFJz9)). I know the projects run in different environments (BPjs is server-side only) but it might still be useful to see how it was implemented. The full repo is here: [https://github.com/bThink-BGU/BPjs](https://t.co/OMX6OiaMQq),

I hope this helps, and will be happy to keep this discussion going. BTW, I'm working on a BP-based tool for QA and product definition ([https://provengo.tech](https://t.co/GvNq8WGFS2)). If you have time to look at it, I'll be happy to know what you think.

Thanks,

-- Michael

**Reply by michbarsinai** [reply]
Clarification for the above: by "event generator" I don't mean regular JavaScript generator functions. Rather, I mean functions that receive the synchronization statements from all b-threads, and create events that can be selected at the given time. This is different from "classic" BP, where the b-program selects from explicitly requested events only (i.e. event objects created by b-threads or pushed into the b-program's external event queue, if there is one).

**Reply by EdwardIrby** [reply]
Hi Michael,

- I'll correct the difference between suspend and block and clear up some of the BP vs bProgram miss naming. 
- I think you're right on the 2nd tldr; I wrote that poorly and even I'm confused, lol. I'll rework that to say  "Modifying program behavior is as easy as adding another thread to the program".

I'll take a look at your slides. I know my implementation of the algorithm is based on this: https://www.wisdom.weizmann.ac.il/~bprogram/BlocklyJS2/. Which I think is classical behavioral programming as referenced in the algorithm below correct?


![Screenshot 2023-06-01 093155](https://github.com/plaited/plaited/assets/1058725/514c6fb3-fa05-4bc7-a8e2-8a75ac592e76)


### RequestIdiom vs ParameterIdiom
So most developers using JS tend to not use generator functions. Instead they are usually used by libraries like the one Ali and I have created.  When I refactored and modernized the code from the above implementation Ali and I eventually realized we needed to abstract away the generator function to prevent avoidable coding errors. I worked with Ali until we found interfaces we really liked. Using these three utility functions we found we could create rather complex behavioral programs free of common generator function usage related errors.

```ts
import { Detail, RuleSet, RulesFunc } from './types.js'

/**
 * @description
 * creates a behavioral thread from synchronization sets and/or other behavioral threads
 */
export const thread = (...rules: RulesFunc<any>[]): RulesFunc<any> =>
  function* () {
    for (const rule of rules) {
      yield* rule()
    }
  }
/**
 * @description
 * A behavioral thread that loops until and optional callback returns false.This 
 * function returns a behavioral thread
 */
export const loop = (
  rules: RulesFunc<any>[],
  condition = () => true
): RulesFunc<any> =>
  function* () {
    while (condition()) {
      for (const rule of rules) {
        yield* rule()
      }
    }
  }
/**
 * @description
 * At synchronization points, each behavioral thread specifies three sets of events:
 * requested events: the threads proposes that these be considered for triggering,
 * and asks to be notified when any of them occurs; waitFor events: the threads does 
 * not request these, but asks to be notified when any of them is triggered; and 
 * blocked events: the threads currently forbids triggering any of these events.
 */
export const sync = <T extends Detail>(
  set: RuleSet<T>
): RulesFunc<T> =>
    function* () {
      yield set
    }

```

In the above `sync` utility function you can see we are always yielding a `RuleSet` object with the following type:

```ts
export type RuleSet<
  T extends (Detail) = (Detail),
> = {
  waitFor?: ParameterIdiom<T> | ParameterIdiom<T>[];
  request?: RequestIdiom<T> | RequestIdiom<T>[];
  block?: ParameterIdiom<T> | ParameterIdiom<T>[];
};
```

Note that we can also pass it an array/list of multiple idioms which is useful when using event selection strategies other than priority queue.

The `RequestIdiom` allows for the passing of data via `detail` which is consumed by feedback actions. However allowing a callback on the `request` introduces unexpected behavior when the `bProgram` runs. 

We're treating the `ParameterIdiom` types like events sets I believe. I found the callback mechanism used in the original implementation for `bp.event` useful for doing things like: 

> any event whose name starts with 'B'

So I made it part of of the type for `ParameterIdiom` which only applies to `waitFor` and `block`. This way you're not just checking the event `type` string identifier but can actually use the full breadth of JS to return back a synchronous function that returns true or false. 

The callback receives the event type and an optional detail object from triggers/requests as arguments. Further when writing test I can add a thread to a bProgram I'm testing that makes use of this callback feature to trigger assertions or `sinon.spy` functions for example thus verifying a thread was called and sync point reached via our feedback actions.

### Event generators

I'm not sure if this is what you mean? But consider the following two examples.

_basic water flow control_
```ts

test('interleave', () => {
  const actual: string[] = []
  const { addThreads, thread, sync, trigger, feedback, loop } = bProgram()
  addThreads({
    addHot: thread(
      sync({ request: { type: 'hot' } }),
      sync({ request: { type: 'hot' } }),
      sync({ request: { type: 'hot' } })
    ),
    addCold: thread(
      sync({ request: { type: 'cold' } }),
      sync({ request: { type: 'cold' } }),
      sync({ request: { type: 'cold' } })
    ),
    mixHotCold: loop([
      sync({
        waitFor: { type: 'hot' },
        block: { type: 'cold' },
      }),
      sync({
        waitFor: { type: 'cold' },
        block: { type: 'hot' },
      }),
    ]),
  })
  feedback({
    hot() {
      actual.push('hot')
    },
    cold() {
      actual.push('cold')
    },
  })
  trigger({ type: 'start' })
  expect(actual).toEqual([
    'hot',
    'cold',
    'hot',
    'cold',
    'hot',
    'cold',
  ])
})
```

_generated water flow control_
```ts
test('interleave', () => {
  const actual: string[] = []
  const { addThreads, thread, sync, trigger, feedback, loop } = bProgram()
  addThreads({
    addHotOnce: thread(
      sync({ request: { type: 'hot_1' } })
    ),
    mixHotCold: loop([
      sync({
        waitFor: { cb: ({ type }) => type.startsWith('hot') },
        block: { cb: ({ type }) => type.startsWith('cold') },
      }),
      sync({
        waitFor: { cb: ({ type }) => type.startsWith('cold')  },
        block: { cb: ({ type }) => type.startsWith('hot') },
      }),
    ]),
  })
  feedback({
    hot_1() {
      actual.push('hot')
      trigger({ type: 'cold' })
      addThreads({
        addMoreHot: thread(
          sync({ request: { type: 'hot' } }),
          sync({ request: { type: 'hot' } })
        ),
        addMoreCold: thread(
          sync({ request: { type: 'cold' } }),
          sync({ request: { type: 'cold' } })
        ),
      })
    },
    cold() {
      actual.push('cold')
    },
    hot() {
      actual.push('hot')
    },
  })
  trigger({ type: 'start' })
  expect(actual).toEqual([
    'hot',
    'cold',
    'hot',
    'cold',
    'hot',
    'cold',
  ])
})
```

Note I'm using the callback in the `waitFor` and `block` idioms to check if the event type start's with hot or cold. I'm also triggering cold in the hot_1 feedback action and then adding more threads with requests to the running program.


### Selection Strategies

I also have two other selection strategies I ship with the library one that randomizes the RequestIdiom's during event selection without effecting the priority queue. The other allows for chaos engineering by ignoring the priority queue and selecting a truly random request. 

I also allow passing a custom eventSelection strategy when invoking the bProgram. My goal was always to see if I could combine the selection strategy with reinforcement learning. I have a prototype I'm working on with tensorflow js.

_selection strategies_
```ts
/** @description Randomized Priority Queue Selection Strategy */
export const randomizedStrategy: Strategy = filteredEvents => {
  for (let i = filteredEvents.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [ filteredEvents[i], filteredEvents[j] ] = [
      filteredEvents[j],
      filteredEvents[i],
    ]
  }
  return filteredEvents.sort(
    ({ priority: priorityA }, { priority: priorityB }) => priorityA - priorityB
  )[0]
}

/** @description Chaos Selection Strategy */
export const chaosStrategy: Strategy = filteredEvents =>
  filteredEvents[Math.floor(Math.random() * Math.floor(filteredEvents.length))]

/** @description Priority Queue Selection Strategy */
export const priorityStrategy: Strategy = filteredEvents =>
  filteredEvents.sort(
    ({ priority: priorityA }, { priority: priorityB }) => priorityA - priorityB
  )[0]

export const selectionStrategies = {
  [strategies.priority]: priorityStrategy,
  [strategies.chaos]: chaosStrategy,
  [strategies.randomized]: randomizedStrategy,
}
```

A user of the library can use different selection strategies likes so passing in `chaos` | `randomized` | `priority` _default_ | or a function of the type `Strategy`.

_Example_
```ts
import { bProgram } from '@plaited/behavioral'

const {...} = bProgram({ strategy: 'randomized'})

```
_Strategy type_
```ts

export type Strategy = (
  filteredEvents: CandidateBid[] | never[],
) => CandidateBid | undefined;
```

**Reply by michbarsinai** [reply]
The algorithm you mention seems correct, and I agree that `ParameterIdiom` are similar to `EventSet`s.

As for generated events - It's a feature people have been talking about, but I haven't seen it implemented yet. Consider the following example of a roving robot. There are three b-threads, two represent sensors that detect obstacles and driving conditions, and one that represents a navigation algorithm. Events are instruction to the robot's engine, and contain angle (steering direction) and motor power. Here are their synchronization bids/statements:

```Javascript

sensor1: {
    block: event => event.angle > 10 && event.angle < 20; // block movement to 10-20 deg., to avoid an obstacle detected in this direction.
}

sensor2: {
   block: event => event.power > 0.5; // e.g. snow mode - block any fast movement that will cause losing traction.
}

nav: {
   request: new EventTemplate({angle:{min:5, max:18}, power:{min:.2, max:.8} })
}
```

Note that `nav` does not request a specific event, but rather requests that an event that meets certain criteria will be requested. The event selection strategy takes all submitted constraints, and requests the following event:

```javascript
{ angle:9, power:.4 }
```

This event was not requested by any b-thread! It was generated by the event selection strategy to meet all constraints posed by the b-threads during the given synchronization point.

So, a cool feature, but not anything I'd lose sleep about ;-)

**Comment by EdwardIrby** [comment]
@michbarsinai if you haven't yet, take a look at the core of our implementation: https://github.com/plaited/plaited/blob/main/libs/behavioral/src/b-program.ts

After looking at your slides I realized we had some similar features such as the snapshots I find those super useful when debugging components, using `bProgram({ dev: (msg) => console.table(msg)} )`
